packages/protocol/contracts-0.8/common/EpochManager.sol

```diff
@@ -506,4 +649,82 @@ contract EpochManager is
       CELOequivalent
     );
   }
+
+  /**
+   * @notice Updates the list of elected validator signers.
+   */
+  function _setElectedSigners(address[] memory _elected) internal {
+    require(electedAccounts.length > 0, "Elected list length cannot be zero.");
+    IAccounts accounts = getAccounts();
+    electedSigners = new address[](_elected.length);
+    for (uint i = 0; i < _elected.length; i++) {
+      electedSigners[i] = accounts.getValidatorSigner(_elected[i]);
+    }
+  }
+
+  /**
+   * @notice Returns the epoch info of a specified blockNumber.
+   * @param _blockNumber Block number of the epoch info is retreived.
+   * @return firstEpoch The first block of the given block number.
+   * @return lastBlock The first block of the given block number.
+   * @return startTimestamp The starting timestamp of the given block number.
+   * @return rewardsBlock The reward block of the given block number.
+   */
+  function _getEpochByBlockNumber(
+    uint256 _blockNumber
+  )
+    internal
+    view
+    onlySystemAlreadyInitialized
+    returns (uint256, uint256, uint256, uint256, uint256)
+  {
+    require(_blockNumber <= block.number, "Invalid blockNumber. Value too high.");
+
+    (uint256 _firstBlockOfFirstEpoch, , , ) = getEpochByNumber(firstKnownEpoch);
+
+    require(_blockNumber >= _firstBlockOfFirstEpoch, "Invalid blockNumber. Value too low.");
+
+    uint256 _firstBlockOfCurrentEpoch = epochs[currentEpochNumber].firstBlock;
+
+    if (_blockNumber >= _firstBlockOfCurrentEpoch) {
+      (
+        uint256 _firstBlock,
+        uint256 _lastBlock,
+        uint256 _startTimestamp,
+        uint256 _rewardsBlock
+      ) = getEpochByNumber(currentEpochNumber);
+      return (currentEpochNumber, _firstBlock, _lastBlock, _startTimestamp, _rewardsBlock);
+    }
+
+    uint256 left = firstKnownEpoch;
+    uint256 right = currentEpochNumber - 1;
+
+    while (left <= right) {
+      uint256 mid = (left + right) / 2;
+      uint256 _epochFirstBlock = epochs[mid].firstBlock;
+      uint256 _epochLastBlock = epochs[mid].lastBlock;
+
+      if (_blockNumber >= _epochFirstBlock && _blockNumber <= _epochLastBlock) {
+        Epoch memory _epoch = epochs[mid];
+        return (
+          mid,
+          _epoch.firstBlock,
+          _epoch.lastBlock,
+          _epoch.startTimestamp,
+          _epoch.rewardsBlock
+        );
+      } else if (_blockNumber < _epochFirstBlock) {
+        right = mid - 1;
+      } else {
+        left = mid + 1;
+      }
+    }
+
+    revert("No matching epoch found for the given block number.");
+  }
```

Comment:
This function is very likely to run out of gas, is it (and the functions that use it) really needed?